File: src/security/policy_engine.ts
/**
 * policy_engine.ts — Phase 4 Policy Enforcement Engine
 *
 * Reads policy.yaml (or falls back to hardcoded defaults) and maps each
 * security pattern match → enforcement action (block / warn / log).
 *
 * Design rules:
 *  - NEVER breaks if policy.yaml is missing or malformed.
 *  - All actions are additive: block always implies warn + log.
 *  - Public API is synchronous to keep the hot scanner path fast.
 */

import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

export type PolicyAction = 'block' | 'warn' | 'log';
export type PolicySeverity = 'low' | 'medium' | 'high' | 'critical';

export interface PolicyRule {
    enabled: boolean;
    action: PolicyAction;
    severity: PolicySeverity;
    description: string;
}

export interface PolicyThresholds {
    pii_density_warn: number;
    pii_density_block: number;
    dataset_risk_score_warn: number;
    prompt_pii_max_items: number;
}

export interface PolicyConfig {
    rules: Record<string, PolicyRule>;
    thresholds: PolicyThresholds;
}

export interface PolicyDecision {
    action: PolicyAction;
    severity: PolicySeverity;
    message: string;
    ruleId: string;
}

// ─────────────────────────────────────────────────────────────────────────────
// Built-in defaults (used when policy.yaml is absent)
// ─────────────────────────────────────────────────────────────────────────────

const DEFAULT_POLICY: PolicyConfig = {
    rules: {
        block_private_keys:      { enabled: true,  action: 'block', severity: 'critical', description: 'Private keys must never appear in workspace.' },
        block_aws_keys:          { enabled: true,  action: 'block', severity: 'critical', description: 'AWS Access Key IDs expose cloud credentials.' },
        block_credit_cards:      { enabled: true,  action: 'block', severity: 'critical', description: 'Credit card numbers are PCI-DSS regulated.' },
        block_connection_strings:{ enabled: true,  action: 'block', severity: 'high',     description: 'Connection strings may embed credentials.' },
        warn_on_api_keys:        { enabled: true,  action: 'warn',  severity: 'high',     description: 'Hard-coded API keys violate secret management.' },
        warn_on_openai_keys:     { enabled: true,  action: 'warn',  severity: 'high',     description: 'OpenAI keys expose paid-API access.' },
        warn_on_jwt:             { enabled: true,  action: 'warn',  severity: 'high',     description: 'JWT tokens grant protected service access.' },
        warn_on_password:        { enabled: true,  action: 'warn',  severity: 'high',     description: 'Hardcoded passwords violate security policy.' },
        warn_on_bearer_token:    { enabled: true,  action: 'warn',  severity: 'high',     description: 'Bearer tokens grant delegated API access.' },
        warn_on_github_token:    { enabled: true,  action: 'warn',  severity: 'high',     description: 'GitHub tokens expose repository access.' },
        warn_on_ssn:             { enabled: true,  action: 'block', severity: 'critical', description: 'SSNs are the highest-risk PII category.' },
        warn_on_email:           { enabled: true,  action: 'warn',  severity: 'medium',   description: 'Email addresses may indicate PII exposure.' },
        warn_on_prompt_pii:      { enabled: true,  action: 'warn',  severity: 'medium',   description: 'PII in LLM prompt — anonymize before sending.' },
        warn_on_prompt_secrets:  { enabled: true,  action: 'block', severity: 'critical', description: 'Secrets in LLM prompts risk third-party exposure.' },
        warn_on_prompt_medical:  { enabled: true,  action: 'warn',  severity: 'high',     description: 'Medical data in prompts may violate HIPAA.' },
        warn_on_high_risk_dataset:{ enabled: true, action: 'warn',  severity: 'high',     description: 'High-risk dataset requires privacy review.' },
        log_dataset_open:        { enabled: true,  action: 'log',   severity: 'low',      description: 'Dataset file opened — logged for audit.' },
    },
    thresholds: {
        pii_density_warn: 0.40,
        pii_density_block: 0.70,
        dataset_risk_score_warn: 60,
        prompt_pii_max_items: 0,
    },
};

// ─────────────────────────────────────────────────────────────────────────────
// Pattern → ruleId mapping
// Maps the pattern name from realtime_scanner to the relevant policy rule key.
// ─────────────────────────────────────────────────────────────────────────────

const PATTERN_TO_RULE: Record<string, string> = {
    'OpenAI API Key':             'warn_on_openai_keys',
    'GitHub Token':               'warn_on_github_token',
    'AWS Access Key':             'block_aws_keys',
    'Generic API Key Assignment': 'warn_on_api_keys',
    'JWT Token':                  'warn_on_jwt',
    'Private Key':                'block_private_keys',
    'Password Assignment':        'warn_on_password',
    'Bearer Token':               'warn_on_bearer_token',
    'Connection String':          'block_connection_strings',
    'Email Address':              'warn_on_email',
    'SSN Pattern':                'warn_on_ssn',
    'Credit Card':                'block_credit_cards',
};

// ─────────────────────────────────────────────────────────────────────────────
// Policy loader
// ─────────────────────────────────────────────────────────────────────────────

let _policy: PolicyConfig = DEFAULT_POLICY;
let _policyLoadedAt: number = 0;
const POLICY_TTL_MS = 30_000; // reload at most every 30 s

function findPolicyFile(extensionPath?: string): string | null {
    // Check workspace root first, then extension dir
    const candidates: string[] = [];
    const wsRoots = vscode.workspace.workspaceFolders?.map(w => w.uri.fsPath) ?? [];
    for (const root of wsRoots) {
        candidates.push(path.join(root, 'policy.yaml'));
    }
    if (extensionPath) {
        candidates.push(path.join(extensionPath, 'policy.yaml'));
    }
    for (const p of candidates) {
        if (fs.existsSync(p)) { return p; }
    }
    return null;
}

/**
 * Simple YAML→object parser for the limited policy.yaml schema.
 * Avoids pulling in a YAML dependency — handles only key: value pairs
 * and nested sections separated by blank lines.
 */
function parseSimpleYaml(text: string): Record<string, any> {
    const result: Record<string, any> = {};
    let section: string | null = null;
    let subSection: string | null = null;

    for (const raw of text.split('\n')) {
        const line = raw.replace(/#.*$/, '').trimEnd();
        if (!line.trim()) { continue; }

        // Detect top-level key (no leading spaces)
        const topMatch = line.match(/^([a-zA-Z_][a-zA-Z0-9_]*):\s*(.*)$/);
        if (topMatch) {
            section = topMatch[1];
            subSection = null;
            if (topMatch[2].trim()) {
                result[section] = coerce(topMatch[2].trim());
            } else {
                result[section] = result[section] ?? {};
            }
            continue;
        }

        // Detect 2-space indented key (sub-section)
        const sub2Match = line.match(/^  ([a-zA-Z_][a-zA-Z0-9_]*):\s*(.*)$/);
        if (sub2Match && section) {
            subSection = sub2Match[1];
            if (typeof result[section] !== 'object') { result[section] = {}; }
            if (sub2Match[2].trim()) {
                (result[section] as any)[subSection] = coerce(sub2Match[2].trim());
            } else {
                (result[section] as any)[subSection] = (result[section] as any)[subSection] ?? {};
            }
            continue;
        }

        // Detect 4-space indented key (leaf values)
        const leaf4Match = line.match(/^    ([a-zA-Z_][a-zA-Z0-9_]*):\s*(.*)$/);
        if (leaf4Match && section && subSection) {
            const key = leaf4Match[1];
            const val = coerce(leaf4Match[2].trim());
            if (typeof (result[section] as any)[subSection] !== 'object') {
                (result[section] as any)[subSection] = {};
            }
            (result[section] as any)[subSection][key] = val;
        }
    }
    return result;
}

function coerce(v: string): boolean | number | string {
    if (v === 'true')  { return true; }
    if (v === 'false') { return false; }
    const n = Number(v);
    if (!isNaN(n) && v !== '') { return n; }
    // Strip surrounding quotes
    return v.replace(/^["']|["']$/g, '');
}

function buildPolicyFromYaml(raw: Record<string, any>): PolicyConfig {
    const rules: PolicyConfig['rules'] = { ...DEFAULT_POLICY.rules };
    const rawRules = raw['rules'] ?? {};
    for (const [id, ruleRaw] of Object.entries(rawRules)) {
        if (!ruleRaw || typeof ruleRaw !== 'object') { continue; }
        const r = ruleRaw as Record<string, any>;
        rules[id] = {
            enabled:     r['enabled']     ?? true,
            action:      (r['action']     ?? 'warn')   as PolicyAction,
            severity:    (r['severity']   ?? 'medium') as PolicySeverity,
            description: r['description'] ?? '',
        };
    }

    const rawThr = raw['thresholds'] ?? {};
    const thresholds: PolicyThresholds = {
        pii_density_warn:       Number(rawThr['pii_density_warn']       ?? DEFAULT_POLICY.thresholds.pii_density_warn),
        pii_density_block:      Number(rawThr['pii_density_block']      ?? DEFAULT_POLICY.thresholds.pii_density_block),
        dataset_risk_score_warn:Number(rawThr['dataset_risk_score_warn'] ?? DEFAULT_POLICY.thresholds.dataset_risk_score_warn),
        prompt_pii_max_items:   Number(rawThr['prompt_pii_max_items']   ?? DEFAULT_POLICY.thresholds.prompt_pii_max_items),
    };
    return { rules, thresholds };
}

function loadPolicy(extensionPath?: string): PolicyConfig {
    const now = Date.now();
    if (now - _policyLoadedAt < POLICY_TTL_MS) { return _policy; }
    _policyLoadedAt = now;

    const filePath = findPolicyFile(extensionPath);
    if (!filePath) { return DEFAULT_POLICY; }

    try {
        const text = fs.readFileSync(filePath, 'utf-8');
        const raw = parseSimpleYaml(text);
        _policy = buildPolicyFromYaml(raw);
    } catch {
        _policy = DEFAULT_POLICY; // graceful fallback
    }
    return _policy;
}

// ─────────────────────────────────────────────────────────────────────────────
// Public API
// ─────────────────────────────────────────────────────────────────────────────

/** Reload the policy immediately (bypasses TTL cache). */
export function refreshPolicy(extensionPath?: string): void {
    _policyLoadedAt = 0;
    loadPolicy(extensionPath);
}

/**
 * Evaluate what action should be taken for a detected pattern.
 *
 * @param patternName  The `name` field from SECURITY_PATTERNS in realtime_scanner
 * @param extensionPath  Extension root (for locating policy.yaml)
 * @returns PolicyDecision or null if the rule is disabled.
 */
export function evaluate(patternName: string, extensionPath?: string): PolicyDecision | null {
    const policy = loadPolicy(extensionPath);
    const ruleId = PATTERN_TO_RULE[patternName];
    if (!ruleId) { return null; }

    const rule = policy.rules[ruleId];
    if (!rule || !rule.enabled) { return null; }

    const emoji = rule.action === 'block' ? '❌' : rule.action === 'warn' ? '⚠' : 'ℹ';
    const label = rule.action === 'block' ? 'blocked by policy' : rule.action === 'warn' ? 'policy warning' : 'logged by policy';

    return {
        action:   rule.action,
        severity: rule.severity,
        ruleId,
        message:  `${emoji} ${patternName} — ${label}: ${rule.description}`,
    };
}

/**
 * Evaluate a dataset risk result against thresholds.
 * Returns a PolicyDecision or null if no threshold is breached.
 */
export function evaluateDataset(
    piiDensity: number,
    riskScore: number,
    extensionPath?: string
): PolicyDecision | null {
    const policy = loadPolicy(extensionPath);
    const thr = policy.thresholds;

    if (piiDensity >= thr.pii_density_block) {
        return {
            action: 'block',
            severity: 'critical',
            ruleId: 'warn_on_high_risk_dataset',
            message: `❌ Dataset PII density ${(piiDensity * 100).toFixed(0)}% exceeds block threshold — blocked by policy`,
        };
    }
    if (piiDensity >= thr.pii_density_warn || riskScore >= thr.dataset_risk_score_warn) {
        const rule = policy.rules['warn_on_high_risk_dataset'];
        if (!rule?.enabled) { return null; }
        return {
            action: 'warn',
            severity: 'high',
            ruleId: 'warn_on_high_risk_dataset',
            message: `⚠ Dataset risk score ${riskScore.toFixed(0)}/100 — policy warning: review before use`,
        };
    }
    return null;
}

/**
 * Evaluate a prompt scan result.
 * Returns a PolicyDecision or null if clean.
 */
export function evaluatePrompt(
    hasCritical: boolean,
    hasHigh: boolean,
    itemCount: number,
    extensionPath?: string
): PolicyDecision | null {
    const policy = loadPolicy(extensionPath);

    if (hasCritical) {
        const rule = policy.rules['warn_on_prompt_secrets'];
        if (rule?.enabled) {
            return {
                action: 'block',
                severity: 'critical',
                ruleId: 'warn_on_prompt_secrets',
                message: `❌ Secrets detected in LLM prompt — blocked by policy: ${rule.description}`,
            };
        }
    }
    if (hasHigh || itemCount > policy.thresholds.prompt_pii_max_items) {
        const rule = policy.rules['warn_on_prompt_pii'];
        if (rule?.enabled) {
            return {
                action: 'warn',
                severity: 'medium',
                ruleId: 'warn_on_prompt_pii',
                message: `⚠ ${itemCount} sensitive item(s) in LLM prompt — policy warning: ${rule.description}`,
            };
        }
    }
    return null;
}

/** Expose thresholds (for dataset monitor). */
export function getThresholds(extensionPath?: string): PolicyThresholds {
    return loadPolicy(extensionPath).thresholds;
}

File: src/security/policy_engine.py
"""
policy_engine.py — Security Policy Engine
==========================================

Supports YAML policy files for data governance rules.

Example policy.yaml:
    rules:
      - block_credit_cards
      - block_medical_records
      - warn_on_api_keys
      - block_ssn
      - warn_on_emails
    
    thresholds:
      max_pii_density: 0.05
      max_risk_score: 50
    
    actions:
      on_violation: "block"  # block | warn | log

The scanner enforces these policies and returns violations.
"""

from __future__ import annotations
import json, os, sys
from typing import Any, Dict, List, Optional
from dataclasses import dataclass, field, asdict

try:
    import yaml
except ImportError as _e:
    raise ImportError(
        "PyYAML is required by policy_engine. Install: pip install pyyaml"
    ) from _e


# ============================================================
# Policy types
# ============================================================

@dataclass
class PolicyViolation:
    rule: str
    severity: str       # "critical", "high", "medium", "low"
    message: str
    action: str         # "block", "warn", "log"
    column: Optional[str] = None
    details: Optional[str] = None


@dataclass
class PolicyResult:
    policy_file: str = ""
    rules_checked: int = 0
    violations: List[Dict[str, Any]] = field(default_factory=list)
    action: str = "pass"  # "block", "warn", "pass"
    summary: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


# ============================================================
# Built-in rules registry
# ============================================================

BUILTIN_RULES: Dict[str, Dict[str, Any]] = {
    "block_credit_cards": {
        "check_type": "pii",
        "category": "credit_card",
        "severity": "critical",
        "action": "block",
        "description": "Block datasets containing credit card numbers"
    },
    "block_ssn": {
        "check_type": "pii",
        "category": "ssn",
        "severity": "critical",
        "action": "block",
        "description": "Block datasets containing SSN patterns"
    },
    "block_medical_records": {
        "check_type": "sensitive",
        "category": "medical_record",
        "severity": "critical",
        "action": "block",
        "description": "Block datasets containing medical records"
    },
    "warn_on_api_keys": {
        "check_type": "secret",
        "category": "api_key_generic",
        "severity": "high",
        "action": "warn",
        "description": "Warn when API keys are detected"
    },
    "warn_on_emails": {
        "check_type": "pii",
        "category": "email",
        "severity": "medium",
        "action": "warn",
        "description": "Warn when email addresses are detected"
    },
    "warn_on_phones": {
        "check_type": "pii",
        "category": "phone_us",
        "severity": "medium",
        "action": "warn",
        "description": "Warn when phone numbers are detected"
    },
    "block_private_keys": {
        "check_type": "secret",
        "category": "private_key",
        "severity": "critical",
        "action": "block",
        "description": "Block content containing private keys"
    },
    "warn_on_passwords": {
        "check_type": "secret",
        "category": "password_field",
        "severity": "high",
        "action": "warn",
        "description": "Warn when password values are detected"
    },
    "block_aws_keys": {
        "check_type": "secret",
        "category": "aws_access_key",
        "severity": "critical",
        "action": "block",
        "description": "Block content containing AWS access keys"
    },
    "block_openai_keys": {
        "check_type": "secret",
        "category": "openai_key",
        "severity": "critical",
        "action": "block",
        "description": "Block content containing OpenAI API keys"
    },
    "warn_on_names": {
        "check_type": "pii",
        "category": "person_name",
        "severity": "medium",
        "action": "warn",
        "description": "Warn when personal names are detected"
    },
    "block_jwt_tokens": {
        "check_type": "secret",
        "category": "jwt_token",
        "severity": "critical",
        "action": "block",
        "description": "Block content containing JWT tokens"
    },
}


# ============================================================
# PolicyEngine
# ============================================================

class PolicyEngine:
    """
    Load and enforce security policies from YAML files.

    Usage:
        engine = PolicyEngine("policy.yaml")
        result = engine.evaluate(scan_report)
    """

    def __init__(self, policy_path: Optional[str] = None):
        self.policy_path = policy_path or ""
        self.rules: List[str] = []
        self.thresholds: Dict[str, Any] = {}
        self.default_action: str = "warn"

        if policy_path and os.path.exists(policy_path):
            self._load_policy(policy_path)
        else:
            # Default policy: all warnings
            self.rules = list(BUILTIN_RULES.keys())

    def _load_policy(self, path: str):
        """Load policy from YAML file."""
        with open(path, "r", encoding="utf-8") as f:
            data = yaml.safe_load(f) or {}

        self.rules = data.get("rules", list(BUILTIN_RULES.keys()))
        self.thresholds = data.get("thresholds", {})
        self.default_action = data.get("actions", {}).get("on_violation", "warn")

    def evaluate(self, scan_report) -> PolicyResult:
        """
        Evaluate a ScanReport against the loaded policy.

        Parameters:
            scan_report: ScanReport from DataScanner

        Returns:
            PolicyResult with violations and recommended action
        """
        result = PolicyResult(
            policy_file=self.policy_path,
            rules_checked=len(self.rules)
        )

        # Build a lookup of findings by (type, category)
        all_findings = []
        for f in scan_report.pii_findings:
            all_findings.append(("pii", f.get("category", ""), f))
        for f in scan_report.secrets:
            all_findings.append(("secret", f.get("category", ""), f))
        for f in scan_report.sensitive_content:
            all_findings.append(("sensitive", f.get("category", ""), f))

        highest_action = "pass"

        for rule_name in self.rules:
            rule_def = BUILTIN_RULES.get(rule_name)
            if not rule_def:
                continue

            check_type = rule_def["check_type"]
            check_cat = rule_def["category"]
            action = rule_def.get("action", self.default_action)

            # Find matching findings
            matches = [f for t, c, f in all_findings if t == check_type and c == check_cat]

            if matches:
                violation = PolicyViolation(
                    rule=rule_name,
                    severity=rule_def["severity"],
                    message=f"{rule_def['description']}: {len(matches)} instance(s) found",
                    action=action,
                    column=matches[0].get("column"),
                    details=f"First found in column '{matches[0].get('column')}'"
                )
                result.violations.append(asdict(violation))

                # Escalate action
                if action == "block":
                    highest_action = "block"
                elif action == "warn" and highest_action != "block":
                    highest_action = "warn"

        # Check threshold violations
        max_risk = self.thresholds.get("max_risk_score", 100)
        if scan_report.risk_score > max_risk:
            result.violations.append(asdict(PolicyViolation(
                rule="risk_score_threshold",
                severity="high",
                message=f"Risk score {scan_report.risk_score:.0f} exceeds threshold {max_risk}",
                action="block"
            )))
            highest_action = "block"

        result.action = highest_action
        n_v = len(result.violations)
        result.summary = (
            f"Checked {result.rules_checked} rules. "
            f"Found {n_v} violation(s). Action: {highest_action.upper()}."
        )

        return result


# ============================================================
# CLI
# ============================================================

@dataclass
class ScanReportProxy:
    """
    Typed proxy for a scan report loaded from JSON in CLI mode.

    Mirrors the fields consumed by PolicyEngine.evaluate() so the CLI
    path uses the same contract as the programmatic path.
    """
    pii_findings:      List[Any] = field(default_factory=list)
    secrets:           List[Any] = field(default_factory=list)
    sensitive_content: List[Any] = field(default_factory=list)
    risk_score:        float      = 0.0


def main(argv=None):
    import argparse
    p = argparse.ArgumentParser(description="Policy Engine — Evaluate scan reports against policy")
    p.add_argument("--policy", default=None, help="Path to policy.yaml")
    p.add_argument("--report", required=True, help="Path to scan report JSON")
    args = p.parse_args(argv)

    with open(args.report, "r", encoding="utf-8") as f:
        report_data = json.load(f)

    report = ScanReportProxy(
        pii_findings      = report_data.get("pii_findings", []),
        secrets           = report_data.get("secrets", []),
        sensitive_content = report_data.get("sensitive_content", []),
        risk_score        = float(report_data.get("risk_score", 0)),
    )

    engine = PolicyEngine(args.policy)
    result = engine.evaluate(report)
    sys.stdout.write(json.dumps(result.to_dict(), indent=2) + "\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
